flang/lib/Lower/CallInterface.cpp

```diff
@@ -31,7 +31,11 @@ bool Fortran::lower::CallerInterface::hasAlternateReturns() const {
 std::string Fortran::lower::CallerInterface::getMangledName() const {
```

Comment:
You are updating `getMangledName` on the callee side, but you should also update the callee side, otherwise, if the BIND(C) function is defined by mean of Fortran, the generated code won't link.

Reply:
Thanks for explanation!
You meant, Caller side here right? Callee side is causing a bit of trouble.
I tried `funit.getSubprogramSymbol()` here https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/lib/Lower/CallInterface.cpp#L168, this is asserting. I'm not sure what's the Canonical way of getting the Symbol ?
Once we get the Symbol this should be fairly straightforward(I think).

Reply:
Yes, sorry, I meant "You are updating getMangledName on the caller side". `getSubprogramSymbol` should be the way to get the symbol, can you provide your example that causes the assert to fire ?

Reply:
Inserting this statement `const auto symbol = funit.getSubprogramSymbol();` in Callee side `getMangledName` leads to following regressions(all of them triggering above assertion):
Should we create a separate issue to capture this(seems non-trivial to me) ?
```
Failed Tests (9):
  Flang :: Examples/hello.f90
  Flang :: Lower/altret.f90
  Flang :: Lower/arithmetic-goto.f90
  Flang :: Lower/assigned-goto.f90
  Flang :: Lower/computed-goto.f90
  Flang :: Lower/format.f90
  Flang :: Lower/io-stmt01.f90
  Flang :: Lower/io-stmt02.f90
  Flang :: Lower/loops.f90

Reply:
I am not sure how you inserted the statement, but note that the previous code was only calling `funit.getSubprogramSymbol()` when `!funit.isMainProgram()`. That is because there is simply no symbol for anymous main programs.
Main programs cannot have BIND attributes anyway, so you should only call `getSubprogramSymbol` when you know you are in a subprogram.
I improved `getSubprogramSymbol` description to help future coders here: https://github.com/flang-compiler/f18-llvm-project/pull/654